Split the allow list and increase the warning level for forever BC breaking changes

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72479
* __->__ #72474

Differential Revision: [D34057168](https://our.internmc.facebook.com/intern/diff/D34057168)

tugsbayasgalan committed:
Split the allow list and increase the warning level for forever BC breaking changes

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34057168).

tugsbayasgalan commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,110 +26,94 @@
-        item[1],
-        re.compile(item[2]) if len(item) > 2 else None,
-    ) for item in ALLOW_LIST if item[1] >= datetime.date.today()
+MAX_ALLOWED_PERIOD = datetime.timedelta(days=365)
```
Rn this is very high until we figure out what we do with the operators that have long future dates

mruberry replied:
I would make this one month and put it above the first list which it applies to, then include in the comment there what this does

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,110 +26,94 @@
+# if you must add an entry to this list.
+ALLOW_LIST_WITHOUT_DEADLINE = [
+    "c10_experimental",
+    "static",
```
There used to be an `Internal` comment line here

tugsbayasgalan replied:
done

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -162,9 +146,12 @@ def check_bc(existing_schemas):
             print("schema: ", str(existing_schema), " found on allowlist, skipping")
             continue
+        if allow_listed_without_deadline(existing_schema):
+            print("schema: {} is found in forever BC breaking list. This is very dangerous".format(str(existing_schema)))
```
Maybe rephrase it to be something like:
{} is in allowlist for BC-breaking evolution without deadline. This is dangerous, do not use unless you are sure there will not be downstream consequences.

tugsbayasgalan replied:
sounds good.

mruberry replied:
We should be clear this is skipping BC and FC schema checks, but I think we would need to add more details about exactly how this is "dangerous" and when, if ever, it's OK to do this. If we're not sure then having someone contact us is probably the appropriate recommendation

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -199,7 +186,7 @@ def check_fc(existing_schemas):
     broken_ops = []
     for existing_schema in existing_schemas:
-        if allow_listed(existing_schema):
+        if allow_listed_with_deadline(existing_schema):
```
I thought we were going to have a separate list for FC breaking allowlist? And why isn't the list without deadline checked here?

tugsbayasgalan replied:
I was gonna do that in a separate PR. In thi PR, I tried not to touch the fc part of the test. But yeah, i think it makes sense to add the one without deadline as well.

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,110 +26,94 @@
-    ("aten::slow_conv_transpose3d", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose3d_backward", datetime.date(2022, 1, 31)),
+ALLOW_LIST_WITH_DEADLINE = [
     ("aten::_svd_helper", datetime.date(2022, 3, 31)),
```
I think all of these can be removed?

tugsbayasgalan replied:
Sounds good

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,110 +26,94 @@
-    ("aten::slow_conv_transpose2d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose3d", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose3d_backward", datetime.date(2022, 1, 31)),
+ALLOW_LIST_WITH_DEADLINE = [
```
Let's extend the comment here to describe what this list is for and how to use it properly (with a good example and a bad example, maybe?)

mruberry replied:
"TEMPORARY_ALLOW_LIST"?

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,110 +26,94 @@
-    ) for item in ALLOW_LIST if item[1] >= datetime.date.today()
+MAX_ALLOWED_PERIOD = datetime.timedelta(days=365)
+
+# WARNING: The operators that are included in this list is extremely unsafe
```
This is a little ungrammatical. Maybe something more like: "Operators included in this list indefinitely bypass all BC and FC schema checks. This is almost certainly NOT what you want to do. See note above."?

tugsbayasgalan replied:
Changed

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,110 +26,94 @@
+# because they would bypass all BC/FC checks for indefinite amount of time.
+# This could cause many severe production issues. Please proceed with extreme caution,
+# if you must add an entry to this list.
+ALLOW_LIST_WITHOUT_DEADLINE = [
```
"INDEFINITE_ALLOW_LIST"?

tugsbayasgalan replied:
Changed

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,110 +26,94 @@
+        if deadline > today and deadline - today < MAX_ALLOWED_PERIOD:
+            output.append((re.compile(item[0]), deadline, re.compile(item[2]) if len(item) > 2 else None))
+        if deadline - today >= MAX_ALLOWED_PERIOD:
+            print("{} will be BC broken for too long. We only allow {} days"
```
This is misleading. The operators in question are not necessarily "broken" for this period, they are just scheming schema BC and FC checks.

This warning should also provide guidance about how the test works and what is an appropriate period to set (not just what is allowed). For example, "Operator foo will skip BC and FC schema checks for X days, but only Y days of skipping the checks are permitted. It's recommended that the skip date be chosen only to permit the PR to be merged. Once the PR is merged for a couple days the operator no longer needs to skip these checks."

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -232,7 +219,7 @@ def check_fc(existing_schemas):
         print("Found forward compatible schemas for all existing schemas")
     else:
-        warnings.warn(
+        print(
```
Should this be upgraded to a failure? It's hard to spot prints and warnings.

The error message here is a little goofy. The first sentence is good, but what does it mean to "contact the PyTorch team"? The operators are also listed as "broken" but they are not necessarily broken, they just a potential forward compatibility schema mismatch. We should try to be as clear as possible in our language so developers can understand the issues our test suite presents to them.

tugsbayasgalan replied:
Changed

tugsbayasgalan committed:
Update on "Split the allow list and increase the warning level for forever BC breaking changes"


Differential Revision: [D34057168](https://our.internmc.facebook.com/intern/diff/D34057168)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34057168).

tugsbayasgalan committed:
Update on "Split the allow list and increase the warning level for forever BC breaking changes"


Differential Revision: [D34057168](https://our.internmc.facebook.com/intern/diff/D34057168)

[ghstack-poisoned]

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,111 +32,84 @@
+        today = datetime.date.today()
+        interval = deadline - today
+        if deadline > today and deadline - today < MAX_ALLOWED_PERIOD:
+            output.append((re.compile(item[0]), deadline, re.compile(item[2]) if len(item) > 2 else None))
```
What is `item[2]`?

tugsbayasgalan replied:
This was here before. (https://github.com/pytorch/pytorch/pull/72474/files#diff-236fbde71e59cb1597cac177a83e49fb62b30770eec55c4e7a0f2650b9eb6203L25) And this is pretty much a copy of what we had before. But now I think about it, it is probably ok to delete it

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -26,111 +32,84 @@
 
+def indefinite_allow_listed(schema):
+    for item in INDEFINITE_ALLOW_LIST:
+        if re.compile(item).search(str(schema)):
```
Maybe we should pre-compile re in `INDEFINITE_ALLOW_LIST` too?

tugsbayasgalan replied:
Why?

tugsbayasgalan committed:
Update on "Split the allow list and increase the warning level for forever BC breaking changes"


Differential Revision: [D34057168](https://our.internmc.facebook.com/intern/diff/D34057168)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34057168).

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -8,127 +8,105 @@
+MAX_ALLOWED_PERIOD = datetime.timedelta(days=30)
 
 # The date specifies how long the allowlist exclusion should apply to.
+# You should pick a date that is far enough in the future that you
```
Nice comment; this is what I was mistakenly looking for in the next PR

tugsbayasgalan committed:
Update on "Split the allow list and increase the warning level for forever BC breaking changes"


Differential Revision: [D34057168](https://our.internmc.facebook.com/intern/diff/D34057168)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34057168).

tugsbayasgalan committed:
Update on "Split the allow list and increase the warning level for forever BC breaking changes"


Differential Revision: [D34057168](https://our.internmc.facebook.com/intern/diff/D34057168)

[ghstack-poisoned]
